### ISSUE TITLE
rook can end up resolving files incorrectly, e.g. appending a URL to a local directory

### ISSUE THREAD

**Issue by agstephens** [roocs/rook#91]
Code appends a file URL to a local directory: '/usr/local/src/rook/https:/data.mips.copernicus-climate.eu/thredds/fileServer/esg_c3s-cmip6/CMIP6/CMIP/IPSL/IPSL-CM6A-LR/historical/r1i1p1f1/Amon/rlds/gr/v20180803/rlds_Amon_IPSL-CM6A-LR_historical_r1i1p1f1_gr_185001-201412.nc'

On delving deep into this, @ellesmith88 and @agstephens think the issue relates to this code:

https://github.com/roocs/rook/blob/master/rook/operator.py#L30-L36

It relates to this issue that is outstanding: https://github.com/roocs/rook/issues/84

However, we don't think that #84 is sensible solution because it requires mapping N inputs to M outputs (i.e. it suggests we can consolidate multiple heterogeneous inputs (dataset IDs, URLs, file paths, glob patterns) into a set of file-paths/glob patterns (that might be of a different length)). We think that will be hard to implement without bugs.

So, we think the solution is to fix this in rook, within `Operator.call()`, with something like:

```
if inputs_are_files(...):
    # This is the case when the inputs are files generated by a previous stage in the workflow
    if they are URLs:
         map them to local file paths (based on the Data Node URL base used in C3S)
         file_paths = mapped_urls(URLs)

    if any of the files do not exist on the file system:
         raise Exception(files expected but not found)
    
    if all files do NOT share the same directory:
         raise Exception(file inputs not in same directory so cannot be processed)

    Need to create a minimum glob-pattern to access only the files that are listed...
    ...because there might be more unrequired files in the directory.

    ...maybe the returned glob is: 
    - "/badc/cmip6/data/.../foo/baa/{full_file_name_1.nc,full_file_name_2.nc}"

    from .input_utils import resolve_collection_if_files

    kwargs = deepcopy(args)
    kwargs['collection'] = resolve_collection_if_files(args.get("collection"))

    call the runner(kwargs)

else:
    # This is the first stage of a workflow so we need to use the Director to test against the inventory
    ...as is currently in place...

```

**Comment by cehbrecht** [comment]
> Code appends a file URL to a local directory: '/usr/local/src/rook/https:/data.mips.copernicus-climate.eu/thredds/fileServer/esg_c3s-cmip6/CMIP6/CMIP/IPSL/IPSL-CM6A-LR/historical/r1i1p1f1/Amon/rlds/gr/v20180803/rlds_Amon_IPSL-CM6A-LR_historical_r1i1p1f1_gr_185001-201412.nc'

I think this happened because the URL (http) was set as a file uri in `build_metalink` (expecting path `/data/test.nc`). See PR #96.

**Comment by agstephens** [comment]
Proposed expansion of ideas in `Operator.call()`:

```

...somewhere...maybe in `roocs-utils` or in `rook`


class FileMapper:

    def __init__(self, file_list, dirpath=None):
          self.dirpath = dirpath
          self.file_list = file_list

          self._resolve()

    def _resolve(self):
         if no dirpath: 
              - check file_list have the same directory or raise exception
              - set self.dirpath to that directory
 
         self.file_list = [os.path.basename(fpath) for fpath in self.file_list]
         self.file_paths = [os.path.join(self.dirpath, fname) for fname in self.file_list]

        loop through and check all files exist in self.file_paths or raise Exception files missing


def is_file_list(collection):
     take first item
     if starts with "/" or starts with "http": return True
    
     return False
     

def url_to_file_path(item):
     NOTE: Use `data_node_root` values to match to item in order to work out file_path,
     E.g.: https://data.mips.copernicus-climate.eu/thredds/fileServer/esg_c3s-cmip6/
     Can take out of DatasetMapper to independent function
     raise not found or return file_path


def map_to_file_paths(collection):
     if mixed collection (i.e. some file path and some URLs): raise not supported
     if first is URL - and so all are URLs:
        try:
          file_paths = [url_to_filepath(item) for item in collection]
         except:
           raise cannot map
    else:
        file_paths = collection

    return file_paths


class Operator(object):
    # Sub-classes require "prefix" property
    prefix = NotImplemented

    def __init__(self, output_dir, apply_fixes=True):
        self.config = {
            "output_dir": output_dir,
        }

    def call(self, args):
        args.update(self.config)
        args['output_dir'] = self._get_output_dir()
        collection = args['collection']

        # Quick hack to find out if collection is a list of files
        runner = self._get_runner()

        if is_file_list(collection):
            # This block is called if this is NOT the first stage of a workflow, and
            # the collection will be a file list (one or more files)
            file_paths = map_to_file_paths(collection)
            file_mapper = FileMapper(file_paths)
            output_uris = runner(args)

        else:
            # This block is called when this is the first stage of a workflow
            director = wrap_director(collection, args, runner)
            output_uris = director.output_uris

        return output_uris

```

**Comment by agstephens** [comment]
@ellesmith88 is implementing this in `Resolve urls` branches/PRs.

**Comment by ellesmith88** [comment]
closed by pr #97
